org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/CompletionProposalReplacementProvider.java

```diff
@@ -166,6 +168,55 @@ public void updateReplacement(CompletionProposal proposal, CompletionItem item,
 	/**
 	 * @param completionBuffer
 	 * @param proposal
+	 * @param range
+	 */
+	private void appendAnonymousClass(StringBuilder completionBuffer, CompletionProposal proposal, Range range) {
+		IDocument document;
+		try {
+			IBuffer buffer = this.compilationUnit.getBuffer();
+			document = JsonRpcHelpers.toDocument(buffer);
+			char[] declarationKey = proposal.getDeclarationKey();
+			if (declarationKey == null) {
+				return;
+			}
+			IJavaProject javaProject = this.compilationUnit.getJavaProject();
+			IJavaElement element = javaProject.findElement(new String(declarationKey), null);
+			if (!(element instanceof IType)) {
+				return;
+			}
+			IType type = (IType) element;
+			int offset = proposal.getReplaceStart();
+			AnonymousTypeCompletionProposal overrider = new AnonymousTypeCompletionProposal(compilationUnit, offset, type, String.valueOf(proposal.getDeclarationSignature()), client.isCompletionSnippetsSupported());
+			String replacement = overrider.updateReplacementString(document, offset, importRewrite);
+			if (proposal.getKind() == CompletionProposal.ANONYMOUS_CLASS_DECLARATION) {
+				range.getStart().setCharacter(range.getStart().getCharacter() - 1);
+				range.getEnd().setCharacter(range.getEnd().getCharacter() + 1);
+			}
+			if (proposal.getKind() == CompletionProposal.ANONYMOUS_CLASS_CONSTRUCTOR_INVOCATION && document.getLength() > offset && range != null) {
+				int pos = -1;
+				if (document.getChar(offset) == '(') {
+					pos = offset;
+				}
+				if (document.getChar(offset + 1) == '(') {
```

Comment:
I can't tell if it should test for `)` or it's `(` a 2nd time, on purpose. Either fix it or document it please.

Reply:
It's correct. I have added a comment.

Reply:
if offset ==  document.getLength()-1, then document.getChar(offset + 1) will throw a BadLocationException.
Although it's unlikely to happen, it still needs to be guarded against

Reply:
it should be 
```java
if (document.getChar(offset + 1) == '(') {
   pos = offset + 1;
} else if (document.getChar(offset) == '(') {
   pos = offset;
]
```

But it's

Reply:
I'm still having a hard time reconciling this piece of code from org.eclipse.jdt.internal.ui.text.java.AnonymousTypeCompletionProposal 
```java
if (pos < document.getLength() && document.getChar(pos) == ')') {
   int currentLength= getReplacementLength();
   if (replacementStringEndsWithParentheses)
     setReplacementLength(currentLength + pos - offset);
   else
     setReplacementLength(currentLength + pos - offset + 1);
}
```
I don't see your comment